### PR TITLE
filterx: fix use after free in `strptime()` and macro caching

### PR DESCRIPTION
**Pull request by alltilla**

### CODE REVIEW

File: lib/filterx/object-datetime.c [hunk 1]
```diff
@@ -255,10 +255,10 @@ _strptime_eval(FilterXExpr *s)
   const gchar *time_str;
   gsize time_str_len;
   gboolean extract_success = filterx_object_extract_string(time_str_obj, &time_str, &time_str_len);
-  filterx_object_unref(time_str_obj);
 
   if (!extract_success)
     {
+      filterx_object_unref(time_str_obj);
```
Comment (MrAnno) [inline lib/filterx/object-datetime.c:261]: We have a lot of `filterx_object_extract_*()` functions, the naming for most of them seems okay, except for `string`, `protobuf`, and `bytes`. For those types, the extracted value is not really extracted but borrowed.
That may have been one of the reasons why this use-after-free was introduced.

I would rename these functions to denote this difference. If we insist on keeping the `extractor` idea in the name, adding `ref` may work too, for example, `filterx_object_extract_string_ref()`.

**Review by MrAnno — COMMENTED** [review]
Nice catch!

File: lib/filterx/object-datetime.c (hunk 1 above)
Comment (OverOrion) [reply lib/filterx/object-datetime.c:261]: Just chiming in, how about `filterx_object_as_string()`?
(Inspired by [Rust conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv))

Comment (alltilla) [reply lib/filterx/object-datetime.c:261]: Good idea. Added `ref` to several getter functions that have this behavior.

Comment (alltilla) [reply lib/filterx/object-datetime.c:261]: To me `filterx_object_as_string()` still does not communicate clearly that it is a borrowed reference to the underlying string. Somehow we should incorporate `ref` to the function name.

Comment (alltilla) [reply lib/filterx/object-datetime.c:261]: Oh my bad the "as" word communicates it.

**Review by MrAnno — APPROVED** [review]

**Pull request merged**